sound/soc/sof/pcm.c

```diff
@@ -370,14 +370,43 @@ static int sof_pcm_trigger(struct snd_pcm_substream *substream, int cmd)
 		return -EINVAL;
 	}
 
-	snd_sof_pcm_platform_trigger(sdev, substream, cmd);
+	/* DMA and IPC sequence is different for start and stop. Need to stop
+	 * DMA before STOP IPC sent.
+	 */
+	switch (cmd) {
+	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
+	case SNDRV_PCM_TRIGGER_RESUME:
+	case SNDRV_PCM_TRIGGER_START:
+		/* start DMA */
+		ret = snd_sof_pcm_platform_trigger(sdev, substream, cmd);
+		if (ret < 0)
+			return ret;
 
-	/* send IPC to the DSP */
-	ret = sof_ipc_tx_message(sdev->ipc, stream.hdr.cmd, &stream,
-				 sizeof(stream), &reply, sizeof(reply));
+		/* send START IPC to the DSP */
+		ret = sof_ipc_tx_message(sdev->ipc, stream.hdr.cmd, &stream,
+					 sizeof(stream), &reply, sizeof(reply));
+		break;
+	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
+	case SNDRV_PCM_TRIGGER_SUSPEND:
+	case SNDRV_PCM_TRIGGER_STOP:
+		/* send STOP IPC to the DSP */
+		ret = sof_ipc_tx_message(sdev->ipc, stream.hdr.cmd, &stream,
+					 sizeof(stream), &reply, sizeof(reply));
+		if (ret < 0)
+			return ret;
+
+		/* stop DMA */
+		ret = snd_sof_pcm_platform_trigger(sdev, substream, cmd);
+		if (ret < 0)
+			return ret;
 
-	if (!ret && reset_hw_params)
-		ret = sof_pcm_dsp_pcm_free(substream, sdev, spcm);
+		if (!ret && reset_hw_params)
+			ret = sof_pcm_dsp_pcm_free(substream, sdev, spcm);
```

Comment:
if the IPC fails this will never be freed. Should it be moved higher before the stop DMA?

Reply:
@xiulipan you could simplify this with a boolean that indicates whether the platform_trigger should be called first or the IPC sent first. ANd this boolean could be set in the first switch statement. Then you can avoid the second switch/case

Reply:
@plbossart @ranj063  
Then we may need a free tag to handle all the errors? As I see we do not check if the `snd_sof_pcm_platform_trigger ` return value.

And we will only free the PCM if stop IPC is success.

Reply:
@xiulipan yes I suppose so. We do not have this situation with the current code but we should free PCM even if IPC fails and then return from the function.